Archives/ClmFile.cpp

```diff
@@ -5,11 +5,8 @@
 #include <stdexcept>
 #include <algorithm>
 
-
 namespace Archives
 {
-	const unsigned int CLM_WRITE_SIZE = 0x00020000;
-
 	const int RIFF = 0x46464952;	// "RIFF"
```

Comment:
I had to move CLM_WRITE_SIZE into the header because it was required in the function prototype of ClmFile::PackFile. I did this to reduce the number of times a new buffer was initialized. It sounds like you wouldn't mind if the buffer was recreated each time PackFile was called. I like this idea better because it simplifies the function prototype of PackFile and allows me to move CLM_WRITE_SIZE back into ClmFile.cpp. I'll go ahead and push this change soon.

I would like to make it read:
```
const uint32_t CLM_WRITE_SIZE = 0x00020000;
const uint32_t RIFF = 0x46464952;	// "RIFF"
const uint32_t WAVE = 0x45564157;	// "WAVE"
const uint32_t FMT  = 0x20746D66;	// "fmt "
const uint32_t DATA = 0x61746164;	// "data"
```

I just don't know if an unsigned value is acceptable for RIFF, WAVE, FMT, and DATA. I'm trying to set all values that will be involved in serialization to a fixed size type to eliminate possible differences between compilers when we get everything cross-platform.

Reply:
Ahh, that makes more sense now.

The RIFF, WAVE, FMT, and DATA values are written to the file as a fixed size, so yes, it makes sense for them to be a fixed sized type, such as uint32_t. Unsigned does make slightly more sense than signed, though it's just a 32-bit ID. I suppose more correct is to have 4 consecutive character codes, making a human readable tag, though I'd only do that if there was a simple and easy to read initialization syntax. Maybe the newer language extensions, or `std::array<char>` better support such initialization.

The CLM_WRITE_SIZE is an array size, so size_t might make the most sense there. That value isn't written to the file as far as I remember. Though I think you pointed out a field that had that value stored in it, so maybe the two should be tied together.

The file copy buffer doesn't need to be fully initialized, since the copy reads data into it first, and then only writes out what was read. However, large arrays on the stack do need to have each page touched to ensure the OS grows the stack properly, and doesn't just cause access violations. Hence, large arrays on the stack need to have at least 1 byte (generally 4 bytes) out of every 4096 bytes touched, which might just be a dummy read. The compiler automatically inserts this code in the function prolog. Hence there is a minor startup delay for functions with a large number of local variables.